File: lib/api/schemas.py
from typing import List
from dataclasses import dataclass

class JSONDeserializable:
    @classmethod
    def from_json(cls, raw: dict | List[dict]) -> 'JSONDeserializable':
        if isinstance(raw, list):
            return [cls(**r) for r in raw]
        return cls(**raw)


@dataclass
class CallSchedule(JSONDeserializable):
    "Моделька для пар"
    timeStart: str
    "Время начала"
    timeEnd: str
    "Время конца"
    number: int
    "Номер"
    length: int
    "Длительность"

@dataclass
class Chair(JSONDeserializable):
    "Моделька для кафедр"
    id: int
    "id кафедры"
    shortName: str
    "Аббревиатура кафедры"
    fullName: str
    "Название кафедры"

@dataclass
class Classroom(JSONDeserializable):
    "Моделька для учебных аудиторий"
    id: int
    "id аудитории"
    name: str
    "Название аудитории"
    countPlace: int
    "Количество мест"
    type: int
    "Тип 0-учебная 1-лаборатория"

@dataclass
class Course(JSONDeserializable):
    "Моделька для курсов"
    course: int
    "Курс"

@dataclass
class Faculty(JSONDeserializable):
    "Моделька для факультетов"
    id: int
    "id факультета"
    shortName: str
    "Аббревиатура факультета(Зависит от языка приложения)"
    fullName: str
    "Название факультета(Зависит от языка приложения)"

@dataclass
class Group(JSONDeserializable):
    "Моделька для групп"
    id: int
    "id группы"
    name: str
    "Название группы"
    course: int
    "Курс группы"
    priority: int
    "Приоритет вывода"
    educationForm: int
    "Форма обучения"

@dataclass
class Person(JSONDeserializable):
    "Моделька для персон"
    id: int
    "id преподавателя"
    firstName: str
    "Имя преподавателя"
    secondName: str
    "Отчество преподавателя"
    lastName: str
    "Фамилия преподавателя"

@dataclass
class Rd(JSONDeserializable):
    "Моделька для объявлений в расписании"
    html: str
    "Объявление"

@dataclass
class Structure(JSONDeserializable):
    "Моделька для структур"
    id: int
    "id структуры"
    shortName: str
    "Аббревиатура структуры(Зависит от языка приложения)"
    fullName: str
    "Название структуры(Зависит от языка приложения)"

@dataclass
class Student(JSONDeserializable):
    "Моделька для Студентов"
    id: int
    "id студента"
    firstName: str
    "Имя студента"
    secondName: str
    "Отчество студента"
    lastName: str
    "Фамилия студента"

@dataclass
class TeacherByName(JSONDeserializable):
    "Моделька для поиска преподавателя"
    chairName: str
    "Название кафедры"
    id: int
    "id преподавателя"
    firstName: str
    "Имя преподавателя"
    secondName: str
    "Отчество преподавателя"
    lastName: str
    "Фамилия преподавателя"

@dataclass
class TimeTablePeriod(JSONDeserializable):
    "Моделька для занятия в расписании"
    r1: int
    "Код занятия"
    rz14: int # По шапке бы надавал за такие названия
    rz15: int
    r5: int
    disciplineId: int
    "Идентификатор глобальной дисциплины"
    educationDisciplineId: int
    "Идентификатор дисциплины учебного плана"
    disciplineFullName: str
    "Полное название дисциплины"
    disciplineShortName: str
    "Сокращенное название дисциплины"
    classroom: str
    "Название аудитории"
    timeStart: str
    "Время начала"
    timeEnd: str
    "Время конца"
    teachersName: str
    "Сокращенные имена преподавателей"
    teachersNameFull: str
    "ФИО преподавателей"
    chairName: str # Missing in the docs
    "Название кафедры"
    type: int
    "Тип занятия"
    typeStr: str # Missing in the docs
    "Тип занятия (строковый формат)"
    dateUpdated: str
    "Дата последнего обновления"
    nonstandardTime: bool
    "Нестандартное время занятия"
    groups: str
    "Названия групп"
    extraText: bool
    "Есть ли объявление"

@dataclass
class TimeTableLesson(JSONDeserializable):
    "Моделька для пары в расписании"
    number: int
    "Номер занятия"
    periods: list[TimeTablePeriod]
    "Список занятий"

    @staticmethod
    def from_json(raw: dict | List[dict]) -> 'TimeTableLesson' | List['TimeTableLesson']:
        if isinstance(raw, list):
            res = []
            for r in raw:
                periods = TimeTablePeriod.from_json(r['periods'])
                del r['periods']
                res.append(TimeTableLesson(**r, periods=periods))
            return res
        periods = TimeTablePeriod.from_json(raw['periods'])
        del raw['periods']
        return TimeTableLesson(**raw, periods=periods)

@dataclass
class TimeTableDate(JSONDeserializable):
    "Моделька для дня в расписании"
    date: str
    "Дата"
    lessons: list[TimeTableLesson]
    "Список пар"

    @staticmethod
    def from_json(raw: dict | List[dict]) -> 'TimeTableDate' | List['TimeTableDate']:
        if isinstance(raw, list):
            res = []
            for r in raw:
                lessons = TimeTableLesson.from_json(r['lessons'])
                del r['lessons']
                res.append(TimeTableDate(**r, lessons=lessons))
            return res
        lessons = TimeTableLesson.from_json(raw['lessons'])
        del raw['lessons']
        return TimeTableDate(**raw, lessons=lessons)

@dataclass
class TeacherIdentifier(JSONDeserializable):
    "Моделька для преподавателей (связь с другими системами)"
    id: int
    "id преподавателя"
    hash: str
    "Hash"

@dataclass
class StudentIdentifier(JSONDeserializable):
    "Моделька для студента (связь с другими системами)"
    id: int
    "id обучения студента"
    hash: str
    "Hash"

@dataclass
class Version(JSONDeserializable):
    "Моделька для версии"
    name: str
    "Версия"
    code: str
    "Код версии"

File: pkg/api/schemas.go
/*
 * Copyright (c) 2022 Bohdan Marukhnenko
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package api

type Structure struct {
	Id        int    `json:"id"`
	ShortName string `json:"shortName"`
	FullName  string `json:"fullName"`
}

type Faculty struct {
	Id        int    `json:"id"`
	ShortName string `json:"shortName"`
	FullName  string `json:"fullName"`
}

type Course struct {
	Course int `json:"course"`
}

type Group struct {
	Id            int    `json:"id"`
	Name          string `json:"name"`
	Course        int    `json:"course"`
	Priority      int    `json:"priority"`
	EducationForm int    `json:"educationForm"`
}

type Student struct {
	Id         int    `json:"id"`
	FirstName  string `json:"firstName"`
	SecondName string `json:"secondName"`
	LastName   string `json:"lastName"`
}

type CallScheduleEntry struct {
	TimeStart string `json:"timeStart"`
	TimeEnd   string `json:"timeEnd"`
	Number    int    `json:"number"`
	Length    int    `json:"length"`
}

type TimeTablePeriod struct {
	R1           int `json:"r1"`
	Rz14         int `json:"rz14"`
	Rz15         int `json:"rz15"`
	R5           int `json:"r5"`
	DisciplineId int `json:"disciplineId"`
	// Disabled because it can sometimes be bool instead of int,
	// which causes errors. University's fault.
	//EducationDisciplineId int    `json:"educationDisciplineId"`
	DisciplineFullName  string `json:"disciplineFullName"`
	DisciplineShortName string `json:"disciplineShortName"`
	Classroom           string `json:"classroom"`
	TimeStart           string `json:"timeStart"`
	TimeEnd             string `json:"timeEnd"`
	TeachersName        string `json:"teachersName"`
	TeachersNameFull    string `json:"teachersNameFull"`
	Type                int    `json:"type"`
	TypeStr             string `json:"typeStr"`
	DateUpdated         string `json:"dateUpdated"`
	NonstandardTime     bool   `json:"nonstandardTime"`
	Groups              string `json:"groups"`
	ChairName           string `json:"chairName"`
	ExtraText           bool   `json:"extraText"`
	LessonYear          int    `json:"lessonYear"`
	Semester            int    `json:"semester"`
}

type TimeTableLesson struct {
	Number  int               `json:"number"`
	Periods []TimeTablePeriod `json:"periods"`
}

type TimeTableDate struct {
	Date    string            `json:"date"`
	Lessons []TimeTableLesson `json:"lessons"`
}

type ScheduleExtraInfo struct {
	Html string `json:"html"`
}

type CallSchedule []CallScheduleEntry
type Schedule []TimeTableDate

func (stud *Student) GetFullName() string {
	return stud.LastName + " " + stud.FirstName + " " + stud.SecondName
}

func (s *CallSchedule) GetCall(number int) *CallScheduleEntry {
	for _, call := range *s {
		if call.Number == number {
			return &call
		}
	}
	return nil
}

func (s *Schedule) GetDay(date string) *TimeTableDate {
	for _, day := range *s {
		if day.Date == date {
			return &day
		}
	}
	return nil
}

// GetLesson returns a lesson with a specific number from a day.
func (day *TimeTableDate) GetLesson(number int) *TimeTableLesson {
	for _, lesson := range day.Lessons {
		if lesson.Number == number {
			return &lesson
		}
	}
	return nil
}

type Chair struct {
	Id        int    `json:"id"`
	ShortName string `json:"shortName"`
	FullName  string `json:"fullName"`
}

type Person struct {
	Id         int    `json:"id"`
	FirstName  string `json:"firstName"`
	SecondName string `json:"secondName"`
	LastName   string `json:"lastName"`
}

func (p *Person) GetFullName() string {
	return p.LastName + " " + p.FirstName + " " + p.SecondName
}
